### PR TITLE
better-retester: persist cache

### PR DESCRIPTION
**Pull request by hongkailiu**
/cc @petr-muller

### CODE REVIEW

File: cmd/better-retester/retester.go [hunk 1]
```diff
@@ -81,10 +135,9 @@ func (c *retestController) sync() error {
 		errs = append(errs, c.retestOrBackoff(pr))
 	}
 
-	// TODO: Load the backoff records from disk after start (using an option) and survive when it is not present
-	// TODO: Persist the backoff records after each sync (atomically, so that we do not corrupt it on interrupt)
```
Comment (hongkailiu) [inline cmd/better-retester/retester.go:85]: atomically: not sure if I did it right.

Comment (hongkailiu) [reply cmd/better-retester/retester.go:85]: Google a bit.
It should be atomic now.

File: cmd/better-retester/retester.go [hunk 2]
```diff
@@ -2,29 +2,95 @@ package main
...
+	}
+	// write to a temp file and rename it to the cache file to ensure "atomic write":
+	// either it is complete or nothing
+	tmpFile, err := ioutil.TempFile("", "backoffCache")
+	if err != nil {
+		return fmt.Errorf("failed to create a temp file: %w", err)
+	}
+	tmp := tmpFile.Name()
+	defer func() {
+		if err := os.Remove(tmp); err != nil {
```
Comment (petr-muller) [inline cmd/better-retester/retester.go:82]: Won't this error out by default when we succeed on `os.Rename` afterwards? After that the `tmp` path won't exist anymore.

File: cmd/better-retester/retester.go [hunk 3]
```diff
@@ -54,11 +121,14 @@ func prUrl(pr tide.PullRequest) string {
 }
 
 func (c *retestController) sync() error {
+	if err := c.backoff.loadFromDisk(); err != nil {
```
[the remark this answers is not in the document]
Comment (hongkailiu) [reply cmd/better-retester/retester.go]: Updated.

File: cmd/better-retester/retester.go (hunk 2 above)
Comment (hongkailiu) [reply cmd/better-retester/retester.go:86]: Updated.

File: cmd/better-retester/retester.go [hunk 4]
```diff
@@ -2,29 +2,95 @@ package main
...
+	bytes, err := ioutil.ReadFile(b.file)
+	if err != nil {
+		return fmt.Errorf("failed to read file %s: %w", b.file, err)
+	}
+	cache := map[string]*PullRequest{}
+	if err := yaml.Unmarshal(bytes, &cache); err != nil {
+		return fmt.Errorf("failed to unmarshal: %w", err)
+	}
+	for key, pr := range cache {
+		if age := metav1.Now().Sub(pr.LastConsideredTime.Time); age > b.cacheRecordAge {
```
[the remark this answers is not in the document]
Comment (hongkailiu) [reply cmd/better-retester/retester.go:56]: Saw the comment on it:

https://github.com/kubernetes/apimachinery/blob/276a8a7530a30f42c82b1f0fa2261a905130bf97/pkg/apis/meta/v1/time.go#L24-L33

Keep the same way to marshal/unmarchal as k8s.
an bad idea?

[the remark this answers is not in the document]
Comment (hongkailiu) [reply cmd/better-retester/retester.go]: Sorry. Misunderstood your question.
Updated.

**Review by petr-muller — APPROVED** [review]
sweet

**Pull request merged**